test/libndctl: Use ndctl_region_set_ro() to change disk read-only state

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=

commit 43e48c0d2f271cba4237f6eefc3e4912a74c102b
Author: Dan Williams <dan.j.williams@intel.com>
Date:   Tue Mar 9 22:09:49 2021 -0800

    test/libndctl: Use ndctl_region_set_ro() to change disk read-only state
    
    Kernel commit 52f019d43c22 ("block: add a hard-readonly flag to struct
    gendisk") broke the read-only management test, by fixing the broken
    behavior that BLKROSET could make a block device read-write even when the
    disk is read-only. The fix [1] propagates changes of the region
    read-only state to the underlying disk. Add ndctl_region_set_ro() ahead of
    BLKROSET so that BLKROSET does not conflict the block_device state with the
    disk state.
    
    [1]: http://lore.kernel.org/r/161534060720.528671.2341213328968989192.stgit@dwillia2-desk3.amr.corp.intel.com
    
    Reported-by: kernel test robot <lkp@intel.com>
    Reported-by: Vishal Verma <vishal.l.verma@intel.com>
    Signed-off-by: Dan Williams <dan.j.williams@intel.com>
    Signed-off-by: Vishal Verma <vishal.l.verma@intel.com>
    Link: https://lore.kernel.org/r/161535658913.530219.12194565167385663385.stgit@dwillia2-desk3.amr.corp.intel.com

Signed-off-by: Jeff Moyer <jmoyer@redhat.com>

diff --git a/test/libndctl.c b/test/libndctl.c
index fc65149..c42f785 100644
--- a/test/libndctl.c
+++ b/test/libndctl.c
@@ -1541,6 +1541,7 @@ static int validate_bdev(const char *devname, struct ndctl_btt *btt,
 		struct ndctl_pfn *pfn, struct ndctl_namespace *ndns,
 		struct namespace *namespace, void *buf)
 {
+	struct ndctl_region *region = ndctl_namespace_get_region(ndns);
 	char bdevpath[50];
 	int fd, rc, ro;
 
@@ -1578,6 +1579,13 @@ static int validate_bdev(const char *devname, struct ndctl_btt *btt,
 	}
 
 	ro = 0;
+	rc = ndctl_region_set_ro(region, ro);
+	if (rc < 0) {
+		fprintf(stderr, "%s: ndctl_region_set_ro failed\n", devname);
+		rc = -errno;
+		goto out;
+	}
+
 	rc = ioctl(fd, BLKROSET, &ro);
 	if (rc < 0) {
 		fprintf(stderr, "%s: BLKROSET failed\n",
@@ -1605,8 +1613,16 @@ static int validate_bdev(const char *devname, struct ndctl_btt *btt,
 		rc = -ENXIO;
 		goto out;
 	}
+
+	rc = ndctl_region_set_ro(region, namespace->ro);
+	if (rc < 0) {
+		fprintf(stderr, "%s: ndctl_region_set_ro reset failed\n", devname);
+		rc = -errno;
+		goto out;
+	}
+
 	rc = 0;
- out:
+out:
 	close(fd);
 	return rc;
 }
